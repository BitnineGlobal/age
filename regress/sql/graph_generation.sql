-- Load Apache AGE extension
LOAD 'age';

-- Set search path to use AGE catalog
SET search_path = ag_catalog;

-- Create a complete graph (5 vertices, 'edges' and 'vertices' labels)
SELECT * FROM create_complete_graph('gp1', 5, 'edges', 'vertices');
SELECT COUNT(*) FROM gp1."edges";
SELECT COUNT(*) FROM gp1."vertices";

-- Query the edges using Cypher
SELECT * FROM cypher('gp1', $$MATCH (a)-[e]->(b) RETURN e$$) AS (n agtype);

-- Drop the graph after use to avoid conflicts
SELECT drop_graph('gp1', true);

-- Test invalid graph creation (expect failure)
-- Attempt to create a graph without edges or vertices labels
SELECT * FROM create_complete_graph('gp3', 5, NULL) -- SHOULD FAIL;
SELECT * FROM create_complete_graph('gp4', NULL, NULL) -- SHOULD FAIL;
SELECT * FROM create_complete_graph(NULL, NULL, NULL) -- SHOULD FAIL;

-- Test for conflicting labels (should error out)
SELECT * FROM create_complete_graph('gp5', 5, 'label', 'label') -- SHOULD FAIL;

-- Drop any remaining test graphs
SELECT drop_graph('gp3', true);
SELECT drop_graph('gp4', true);
SELECT drop_graph('gp5', true);

-- Test barbell graph generation (5 vertices per side, no intermediate vertices)
SELECT * FROM age_create_barbell_graph('gp1', 5, 0, 'vertices', NULL, 'edges', NULL);
SELECT COUNT(*) FROM gp1."edges";
SELECT COUNT(*) FROM gp1."vertices";

-- Run Cypher query to retrieve edges in the barbell graph
SELECT * FROM cypher('gp1', $$MATCH (a)-[e]->(b) RETURN e$$) AS (n agtype);

-- Drop the graph to avoid conflicts
SELECT drop_graph('gp1', true);

-- Test barbell graph with 10 intermediate vertices
SELECT * FROM age_create_barbell_graph('gp2', 5, 10, 'vertices', NULL, 'edges', NULL);
SELECT COUNT(*) FROM gp2."edges";
SELECT COUNT(*) FROM gp2."vertices";

-- Test invalid barbell graph creation (should fail)
SELECT * FROM age_create_barbell_graph(NULL, NULL, NULL, NULL, NULL, NULL, NULL) -- SHOULD FAIL;
SELECT * FROM age_create_barbell_graph('gp2', NULL, 0, 'vertices', NULL, 'edges', NULL) -- SHOULD FAIL;
SELECT * FROM age_create_barbell_graph('gp3', 5, NULL, 'vertices', NULL, 'edges', NULL) -- SHOULD FAIL;
SELECT * FROM age_create_barbell_graph('gp4', NULL, 0, 'vertices', NULL, 'edges', NULL) -- SHOULD FAIL;
SELECT * FROM age_create_barbell_graph('gp5', 5, 0, 'vertices', NULL, NULL, NULL) -- SHOULD FAIL;

-- Test conflicting labels in barbell graph (should error out)
SELECT * FROM age_create_barbell_graph('gp6', 5, 10, 'label', NULL, 'label', NULL) -- SHOULD FAIL;

-- Drop remaining graphs
SELECT drop_graph('gp2', true);
SELECT drop_graph('gp3', true);
SELECT drop_graph('gp4', true);
SELECT drop_graph('gp5', true);
SELECT drop_graph('gp6', true);
