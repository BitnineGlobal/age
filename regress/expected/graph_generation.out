/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
LOAD 'age';
SET search_path = ag_catalog;
SELECT * FROM create_complete_graph('gp1',5,'edges','vertices');
NOTICE:  graph "gp1" has been created
NOTICE:  VLabel "vertices" has been created
NOTICE:  ELabel "edges" has been created
 create_complete_graph 
-----------------------
 
(1 row)

SELECT COUNT(*) FROM gp1."edges";
 count 
-------
    10
(1 row)

SELECT COUNT(*) FROM gp1."vertices";
 count 
-------
     5
(1 row)

SELECT * FROM cypher('gp1', $$MATCH (a)-[e]->(b) RETURN e$$) as (n agtype);
                                                             n                                                              
----------------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842625, "label": "edges", "end_id": 844424930131970, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842629, "label": "edges", "end_id": 844424930131971, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842626, "label": "edges", "end_id": 844424930131971, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842630, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842627, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842632, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842631, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842634, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131972, "properties": {}}::edge
 {"id": 1125899906842633, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842628, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131969, "properties": {}}::edge
(10 rows)

SELECT * FROM create_complete_graph('gp1',5,'edges','vertices');
 create_complete_graph 
-----------------------
 
(1 row)

SELECT COUNT(*) FROM gp1."edges";
 count 
-------
    20
(1 row)

SELECT COUNT(*) FROM gp1."vertices";
 count 
-------
    10
(1 row)

SELECT * FROM create_complete_graph('gp2',5,'edges');
NOTICE:  graph "gp2" has been created
NOTICE:  ELabel "edges" has been created
 create_complete_graph 
-----------------------
 
(1 row)

-- SHOULD FAIL
SELECT * FROM create_complete_graph('gp3',5, NULL);
ERROR:  edge label can not be NULL
SELECT * FROM create_complete_graph('gp4',NULL,NULL);
ERROR:  number of nodes can not be NULL
SELECT * FROM create_complete_graph(NULL,NULL,NULL);
ERROR:  graph name can not be NULL
-- Should error out because same labels are used for both vertices and edges
SELECT * FROM create_complete_graph('gp5',5,'label','label');
ERROR:  vertex and edge label can not be same
-- DROPPING GRAPHS
SELECT drop_graph('gp1', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table gp1._ag_label_vertex
drop cascades to table gp1._ag_label_edge
drop cascades to table gp1.vertices
drop cascades to table gp1.edges
NOTICE:  graph "gp1" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('gp2', true);
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to table gp2._ag_label_vertex
drop cascades to table gp2._ag_label_edge
drop cascades to table gp2.edges
NOTICE:  graph "gp2" has been dropped
 drop_graph 
------------
 
(1 row)

-- Tests for barbell graph generation
SELECT * FROM age_create_barbell_graph('gp1',5,0,'vertices',NULL,'edges',NULL);
NOTICE:  graph "gp1" has been created
NOTICE:  VLabel "vertices" has been created
NOTICE:  ELabel "edges" has been created
 age_create_barbell_graph 
--------------------------
 
(1 row)

SELECT COUNT(*) FROM gp1."edges";
 count 
-------
    21
(1 row)

SELECT COUNT(*) FROM gp1."vertices";
 count 
-------
    10
(1 row)

SELECT * FROM cypher('gp1', $$MATCH (a)-[e]->(b) RETURN e$$) as (n agtype);
                                                             n                                                              
----------------------------------------------------------------------------------------------------------------------------
 {"id": 1125899906842625, "label": "edges", "end_id": 844424930131970, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842629, "label": "edges", "end_id": 844424930131971, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842626, "label": "edges", "end_id": 844424930131971, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842627, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842632, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842630, "label": "edges", "end_id": 844424930131972, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842634, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131972, "properties": {}}::edge
 {"id": 1125899906842628, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131969, "properties": {}}::edge
 {"id": 1125899906842631, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131970, "properties": {}}::edge
 {"id": 1125899906842633, "label": "edges", "end_id": 844424930131973, "start_id": 844424930131971, "properties": {}}::edge
 {"id": 1125899906842635, "label": "edges", "end_id": 844424930131975, "start_id": 844424930131974, "properties": {}}::edge
 {"id": 1125899906842639, "label": "edges", "end_id": 844424930131976, "start_id": 844424930131975, "properties": {}}::edge
 {"id": 1125899906842636, "label": "edges", "end_id": 844424930131976, "start_id": 844424930131974, "properties": {}}::edge
 {"id": 1125899906842637, "label": "edges", "end_id": 844424930131977, "start_id": 844424930131974, "properties": {}}::edge
 {"id": 1125899906842642, "label": "edges", "end_id": 844424930131977, "start_id": 844424930131976, "properties": {}}::edge
 {"id": 1125899906842640, "label": "edges", "end_id": 844424930131977, "start_id": 844424930131975, "properties": {}}::edge
 {"id": 1125899906842644, "label": "edges", "end_id": 844424930131978, "start_id": 844424930131977, "properties": {}}::edge
 {"id": 1125899906842638, "label": "edges", "end_id": 844424930131978, "start_id": 844424930131974, "properties": {}}::edge
 {"id": 1125899906842641, "label": "edges", "end_id": 844424930131978, "start_id": 844424930131975, "properties": {}}::edge
 {"id": 1125899906842643, "label": "edges", "end_id": 844424930131978, "start_id": 844424930131976, "properties": {}}::edge
 {"id": 1125899906842645, "label": "edges", "end_id": 844424930131978, "start_id": 844424930131969, "properties": {}}::edge
(21 rows)

SELECT * FROM age_create_barbell_graph('gp1',5,0,'vertices',NULL,'edges',NULL);
 age_create_barbell_graph 
--------------------------
 
(1 row)

SELECT COUNT(*) FROM gp1."edges";
 count 
-------
    42
(1 row)

SELECT COUNT(*) FROM gp1."vertices";
 count 
-------
    20
(1 row)

SELECT * FROM age_create_barbell_graph('gp2',5,10,'vertices',NULL,'edges',NULL);
NOTICE:  graph "gp2" has been created
NOTICE:  VLabel "vertices" has been created
NOTICE:  ELabel "edges" has been created
 age_create_barbell_graph 
--------------------------
 
(1 row)

-- SHOULD FAIL
SELECT * FROM age_create_barbell_graph(NULL,NULL,NULL,NULL,NULL,NULL,NULL);
ERROR:  Graph name cannot be NULL
SELECT * FROM age_create_barbell_graph('gp2',NULL,0,'vertices',NULL,'edges',NULL); 
ERROR:  Graph size cannot be NULL or lower than 3
SELECT * FROM age_create_barbell_graph('gp3',5,NULL,'vertices',NULL,'edges',NULL);
ERROR:  Bridge size cannot be NULL or lower than 0
SELECT * FROM age_create_barbell_graph('gp4',NULL,0,'vertices',NULL,'edges',NULL);
ERROR:  Graph size cannot be NULL or lower than 3
SELECT * FROM age_create_barbell_graph('gp5',5,0,'vertices',NULL,NULL,NULL);
ERROR:  edge label can not be NULL
-- Should error out because same labels are used for both vertices and edges
SELECT * FROM age_create_barbell_graph('gp6',5,10,'label',NULL,'label',NULL);
ERROR:  vertex and edge label can not be same
-- PATH GRAPH TESTS
-- Should throw an error because number of vertices cannot be less than 2.
SELECT age_create_path('path_graph_1', 1, NULL, NULL, false);
NOTICE:  graph "path_graph_1" has been created
ERROR:  Number of vertices in path cannot be less than 2.
-- Should create 5 vertices with 4 edges. All standard ag_labels.
SELECT age_create_path('path_graph_5_false', 5, NULL, NULL, false);
NOTICE:  graph "path_graph_5_false" has been created
 age_create_path 
-----------------
 
(1 row)

SELECT * FROM path_graph_5_false._ag_label_vertex;
       id        | properties 
-----------------+------------
 281474976710657 | {}
 281474976710658 | {}
 281474976710659 | {}
 281474976710660 | {}
 281474976710661 | {}
(5 rows)

SELECT * FROM path_graph_5_false._ag_label_edge;
       id        |    start_id     |     end_id      | properties 
-----------------+-----------------+-----------------+------------
 562949953421313 | 281474976710660 | 281474976710661 | {}
 562949953421314 | 281474976710659 | 281474976710660 | {}
 562949953421315 | 281474976710658 | 281474976710659 | {}
 562949953421316 | 281474976710657 | 281474976710658 | {}
(4 rows)

-- Should create 5 vertices with 8 edges. All standard ag_labels.
SELECT age_create_path('path_graph_5_true', 5, NULL, NULL, true);
NOTICE:  graph "path_graph_5_true" has been created
 age_create_path 
-----------------
 
(1 row)

SELECT * FROM path_graph_5_true._ag_label_vertex;
       id        | properties 
-----------------+------------
 281474976710657 | {}
 281474976710658 | {}
 281474976710659 | {}
 281474976710660 | {}
 281474976710661 | {}
(5 rows)

SELECT * FROM path_graph_5_true._ag_label_edge;
       id        |    start_id     |     end_id      | properties 
-----------------+-----------------+-----------------+------------
 562949953421313 | 281474976710660 | 281474976710661 | {}
 562949953421313 | 281474976710661 | 281474976710660 | {}
 562949953421314 | 281474976710659 | 281474976710660 | {}
 562949953421314 | 281474976710660 | 281474976710659 | {}
 562949953421315 | 281474976710658 | 281474976710659 | {}
 562949953421315 | 281474976710659 | 281474976710658 | {}
 562949953421316 | 281474976710657 | 281474976710658 | {}
 562949953421316 | 281474976710658 | 281474976710657 | {}
(8 rows)

-- Should create 10 vertices with 9 edges. New ag_labels.
SELECT age_create_path('path_graph_10_false', 10, 'PathVertex', 'PATH_EDGE', false);
NOTICE:  graph "path_graph_10_false" has been created
NOTICE:  VLabel "PathVertex" has been created
NOTICE:  ELabel "PATH_EDGE" has been created
 age_create_path 
-----------------
 
(1 row)

SELECT * FROM path_graph_10_false."PathVertex";
       id        | properties 
-----------------+------------
 844424930131969 | {}
 844424930131970 | {}
 844424930131971 | {}
 844424930131972 | {}
 844424930131973 | {}
 844424930131974 | {}
 844424930131975 | {}
 844424930131976 | {}
 844424930131977 | {}
 844424930131978 | {}
(10 rows)

SELECT * FROM path_graph_10_false."PATH_EDGE";
        id        |    start_id     |     end_id      | properties 
------------------+-----------------+-----------------+------------
 1125899906842625 | 844424930131977 | 844424930131978 | {}
 1125899906842626 | 844424930131976 | 844424930131977 | {}
 1125899906842627 | 844424930131975 | 844424930131976 | {}
 1125899906842628 | 844424930131974 | 844424930131975 | {}
 1125899906842629 | 844424930131973 | 844424930131974 | {}
 1125899906842630 | 844424930131972 | 844424930131973 | {}
 1125899906842631 | 844424930131971 | 844424930131972 | {}
 1125899906842632 | 844424930131970 | 844424930131971 | {}
 1125899906842633 | 844424930131969 | 844424930131970 | {}
(9 rows)

-- Should create 10 vertices with 18 edges. New ag_labels.
SELECT age_create_path('path_graph_10_true', 10, 'PathVertex', 'PATH_EDGE', true);
NOTICE:  graph "path_graph_10_true" has been created
NOTICE:  VLabel "PathVertex" has been created
NOTICE:  ELabel "PATH_EDGE" has been created
 age_create_path 
-----------------
 
(1 row)

SELECT * FROM path_graph_10_true."PathVertex";
       id        | properties 
-----------------+------------
 844424930131969 | {}
 844424930131970 | {}
 844424930131971 | {}
 844424930131972 | {}
 844424930131973 | {}
 844424930131974 | {}
 844424930131975 | {}
 844424930131976 | {}
 844424930131977 | {}
 844424930131978 | {}
(10 rows)

SELECT * FROM path_graph_10_true."PATH_EDGE";
        id        |    start_id     |     end_id      | properties 
------------------+-----------------+-----------------+------------
 1125899906842625 | 844424930131977 | 844424930131978 | {}
 1125899906842625 | 844424930131978 | 844424930131977 | {}
 1125899906842626 | 844424930131976 | 844424930131977 | {}
 1125899906842626 | 844424930131977 | 844424930131976 | {}
 1125899906842627 | 844424930131975 | 844424930131976 | {}
 1125899906842627 | 844424930131976 | 844424930131975 | {}
 1125899906842628 | 844424930131974 | 844424930131975 | {}
 1125899906842628 | 844424930131975 | 844424930131974 | {}
 1125899906842629 | 844424930131973 | 844424930131974 | {}
 1125899906842629 | 844424930131974 | 844424930131973 | {}
 1125899906842630 | 844424930131972 | 844424930131973 | {}
 1125899906842630 | 844424930131973 | 844424930131972 | {}
 1125899906842631 | 844424930131971 | 844424930131972 | {}
 1125899906842631 | 844424930131972 | 844424930131971 | {}
 1125899906842632 | 844424930131970 | 844424930131971 | {}
 1125899906842632 | 844424930131971 | 844424930131970 | {}
 1125899906842633 | 844424930131969 | 844424930131970 | {}
 1125899906842633 | 844424930131970 | 844424930131969 | {}
(18 rows)

-- DROPPING GRAPHS
SELECT drop_graph('gp1', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table gp1._ag_label_vertex
drop cascades to table gp1._ag_label_edge
drop cascades to table gp1.vertices
drop cascades to table gp1.edges
NOTICE:  graph "gp1" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('gp2', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table gp2._ag_label_vertex
drop cascades to table gp2._ag_label_edge
drop cascades to table gp2.vertices
drop cascades to table gp2.edges
NOTICE:  graph "gp2" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('path_graph_5_false', true);
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table path_graph_5_false._ag_label_vertex
drop cascades to table path_graph_5_false._ag_label_edge
NOTICE:  graph "path_graph_5_false" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('path_graph_5_true', true);
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table path_graph_5_true._ag_label_vertex
drop cascades to table path_graph_5_true._ag_label_edge
NOTICE:  graph "path_graph_5_true" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('path_graph_10_false', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table path_graph_10_false._ag_label_vertex
drop cascades to table path_graph_10_false._ag_label_edge
drop cascades to table path_graph_10_false."PathVertex"
drop cascades to table path_graph_10_false."PATH_EDGE"
NOTICE:  graph "path_graph_10_false" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT drop_graph('path_graph_10_true', true);
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table path_graph_10_true._ag_label_vertex
drop cascades to table path_graph_10_true._ag_label_edge
drop cascades to table path_graph_10_true."PathVertex"
drop cascades to table path_graph_10_true."PATH_EDGE"
NOTICE:  graph "path_graph_10_true" has been dropped
 drop_graph 
------------
 
(1 row)

